vermouth/ffinput.py

```diff
@@ -51,196 +55,337 @@
+            method, kwargs = self.METH_DICT[tuple(self.section)]
+        #elif wildcard_section in self.METH_DICT:
+        #    method, kwargs = self.METH_DICT[wildcard_section]
+        else:
+            raise IOError("Can't parse line {} in section '{}' because the "
+                          "section is unknown".format(lineno, self.section))
+        try:
+            return method(self, line, lineno, **kwargs)
+        except Exception as error:
+            raise IOError("Problems parsing line {}. I think it should be a "
+                          "'{}' line, but I can't parse it as such."
+                          "".format(lineno, self.section)) from error
+
+    def finalize_section(self, pervious_section, ended_section):
+
+        """
+        Called once a section is finished. It appends the current_links list
+        to the links and update the block dictionary with current_block. Thereby it
+        finishes the reading a given section. 
+
+        Parameters
+        ---------
+        previous_section: list[str]
+            The last parsed section.
+        ended_section: list[str]
+            The sections that have been ended.
+        """
+
+        # type comparison is what makes this work!
+        
+        if type(self.current_block) != type(None):
+           self.blocks[self.current_block.name] = self.current_block
+
+        if type(self.current_link) != type(None):
+           self.links.append(self.current_link)
+       
+    def finalize(self, lineno=0):
+
+        # super needs to go first because it calls finilze_section
+        # which appends the last links and blocks. If it does not 
+        # the blocks are not updated into the ff directory
+
+        super().finalize(lineno=lineno)
+
+        for block in self.blocks.values():
+            # Because of hos they are described in gromacs, proper and improper
+            # dihedral angles are all under the [ dihedrals ] section. However
+            # the way they are treated differently in the library, at least on how
+            # they generate edges. Here we move the all the impropers into their own
+            # [ impropers ] section.
+            propers = []
+            impropers = []
+            for dihedral in block.interactions.get('dihedrals', []):
+                if dihedral.parameters and dihedral.parameters[0] == '2':
+                    impropers.append(dihedral)
+                else:
+                    propers.append(dihedral)
+            block.interactions['dihedrals'] = propers
+            block.interactions['impropers'] = impropers
+            block.make_edges_from_interactions()
+        for link in self.links:
+            link.make_edges_from_interactions()
+        self.force_field.blocks.update(self.blocks)
+        self.force_field.links.extend(self.links)
+        modifications = {mod.name: mod for mod in self.modifications}
+        self.force_field.modifications.update(modifications)
```

Comment:
Even the self.(blocks|links|modifications) you could remove in favour of just adding them to self.force_field as they're being parsed.

Reply:
this is actually quite nice because it makes the finilize method in ffinput completely obsolete.